server.js

```diff
@@ -1,13 +1,217 @@
+  }
+
+  try {
+    const palettes = await database('palettes').where('project_id', id).select();
+    if(!palettes.length) {
+      return response.status(404).json({ error: `Project with ID of ${id} does not have any palettes` });
+    }
+    return response.status(200).json(palettes);
+  } catch (error) {
+    return response.status(500).json({ error });
+  }
+});
+
+app.get('/api/v1/palettes/:id', async (request, response) => {
+  const { id } = request.params;
+
+  if(!parseInt(id)) {
+    return response.status(422).json({ error: `Incorrect ID: ${id}, Required data type: <Number>`})
+  }
+
+  try { 
+    const palette = await database('palettes').where('id', id).select();
+    if(!palette.length) {
+      return response.status(404).json({ error: `Could not locate palette: ${id}` });
+    }
+    return response.status(200).json(palette);
+  } catch (error) {
+    return response.status(500).json({ error });
+  }
+});
+
+app.post('/api/v1/projects', async (request, response) => {
+  const project = request.body;
+
+  for (let requiredParam of [ 'name' ]) {
+    if (!project[requiredParam]) {
+      return response.status(422).send({ error: `Expected format: { name: <String> }, Your missing a ${[requiredParam]} property`});
+    }
+  }
+
+  try {
+    const id = await database('projects').insert(project, 'id');
+    return response.status(201).json({ id: id[0] });
+  } catch (error) {
+    return response.status(500).json({ error });
+  }
+});
+
+app.post('/api/v1/palettes', async (request, response) => {
+  const palette = request.body;
+
+  for (let requiredParam of [ 'name', 'project_id', 'color_one', 'color_two', 'color_three', 'color_four', 'color_five' ]) {
+    if (!palette[requiredParam]) {
+      return response.status(422).send({ error: `Expected format: { name: <String>, project_id: <Number>, color_one:<String>, color_two:<String>, color_three:<String>, color_four:<String>, color_five:<String>} Your missing a ${[requiredParam]} property`});
+    }
+  }
+
+  try {
+    const id = await database('palettes').insert(palette, 'id');
+    return response.status(201).json({ id: id[0] });
+  } catch (error) {
+    return response.status(500).json({ error });
+  }
+});
+
+app.patch('/api/v1/palettes/:id', async (request, response) => {
+  const { id } = request.params;
+  const patch = request.body;
+  const patchKeys = Object.keys(patch);
+
+  if(!parseInt(id)) {
+    return response.status(422).json({ error: `Incorrect ID: ${id}, Required data type: <Number>`});
+  }
+
+  for (let requiredParam of [ 'color_one', 'color_two', 'color_three', 'color_four', 'color_five' ]) {
+    if (!patch[requiredParam]) {
+      return response.status(422).send({ error: `Expected format: { color_one:<String>, color_two:<String>, color_three:<String>, color_four:<String>, color_five:<String>} Your missing a ${[requiredParam]} property`});
+    }
+  }
+
+  for (let i = 0; i < patchKeys.length; i++) {
```

Comment:
We talked about the possibility of refining this error handling as it ends up being quite a bit. We are checking to make sure that ALL the keys being patched are contained in the body, so you have to patch all colors at once and can not patch just a single or a couple. We are also then verifying that ONLY the color keys are being patched an no other keys are being updated. Not sure if there is a way to refactor this into one loop that checks all that? It might also be just fine the way it is - it is working the way we want it to!